packages/java/parser-jvm-core/src/main/java/dev/hilla/parser/core/AbstractCompositePlugin.java

```diff
@@ -0,0 +1,69 @@
+package dev.hilla.parser.core;
+
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import javax.annotation.Nonnull;
+
+import dev.hilla.parser.utils.PluginException;
+
+public abstract class AbstractCompositePlugin<C extends PluginConfiguration>
+        extends AbstractPlugin<C> {
+    private final LinkedList<Plugin> plugins = new LinkedList<>();
+
+    protected AbstractCompositePlugin(@Nonnull Plugin... plugins) {
+        Stream.of(plugins).map(Objects::requireNonNull)
+                .forEachOrdered(this.plugins::add);
+        verifyPluginsOrder();
+    }
+
+    @Override
+    public void enter(NodePath<?> nodePath) {
+        plugins.iterator().forEachRemaining((plugin) -> plugin.enter(nodePath));
+    }
+
+    @Override
+    public void exit(NodePath<?> nodePath) {
+        plugins.descendingIterator()
+                .forEachRemaining((plugin) -> plugin.exit(nodePath));
+    }
+
+    @Override
+    @Nonnull
+    public NodeDependencies scan(@Nonnull NodeDependencies nodeDependencies) {
+        for (var plugin : plugins) {
+            nodeDependencies = plugin.scan(nodeDependencies);
+        }
+        return nodeDependencies;
+    }
+
+    @Override
+    public void setStorage(SharedStorage storage) {
+        super.setStorage(storage);
+        plugins.iterator()
+                .forEachRemaining(plugin -> plugin.setStorage(storage));
+    }
+
+    private void verifyPluginsOrder() {
```

Comment:
I think that, given that the order is imposed by the new strategy, verifying it is useless